### PR TITLE
added validation of signature without stripping port number

### PR DESCRIPTION
**Pull request by eshanholtz**
PR for DI-319

### CODE REVIEW

File: src/Twilio/Security/RequestValidator.php [hunk 1]
```diff
@@ -67,35 +67,41 @@ public static function computeBodyHash($data = '') {
...
      * @param array $data
-     * @return bool 
+     * @return bool
      */
 
     public function validate($expectedSignature, $url, $data = array()) {
 
-        if (\is_array($data)) {
+        $urlWithPort = RequestValidator::addPort(\parse_url($url));
+        $urlWithoutPort = RequestValidator::removePort(\parse_url($url));
```
Comment (childish-sambino) [inline src/Twilio/Security/RequestValidator.php]: nit: Just do `parse_url` once and store in a local var.

File: src/Twilio/Security/RequestValidator.php [hunk 2]
```diff
@@ -67,35 +67,41 @@ public static function computeBodyHash($data = '') {
...
 
-            return self::compare(
-                $this->computeSignature($url),
-                $expectedSignature
-            ) && self::compare(
-                self::computeBodyHash($data),
-                $params['bodySHA256']
-            );
-
+            $params['bodySHA256'] = (isset($params['bodySHA256'])) ? $params['bodySHA256'] : NULL;
```
Comment (childish-sambino) [inline src/Twilio/Security/RequestValidator.php]: Why is this line needed?

File: src/Twilio/Security/RequestValidator.php [hunk 3]
```diff
@@ -67,35 +67,41 @@ public static function computeBodyHash($data = '') {
...
+            $validBodyHash = self::compare(self::computeBodyHash($data), $params['bodySHA256']);
+            $data = array();
         }
 
+        /*  Check signature of the URL with and without port information
```
Comment (childish-sambino) [inline src/Twilio/Security/RequestValidator.php]: nit: Start block comment text on next line.

```
/*
 * Comment.
 */
```

File: src/Twilio/Security/RequestValidator.php [hunk 4]
```diff
@@ -67,35 +67,41 @@ public static function computeBodyHash($data = '') {
...
+         *  since sig generation on the back end is inconsistent.
+         */
+        $validSignatureWithPort = self::compare(
+            $this->computeSignature($urlWithPort, $data),
+            $expectedSignature
+        );
+        $validSignatureWithoutPort = self::compare(
+            $this->computeSignature($urlWithoutPort, $data),
+            $expectedSignature
+        );
```
Comment (childish-sambino) [inline src/Twilio/Security/RequestValidator.php]: Nice simplification!

File: src/Twilio/Security/RequestValidator.php [hunk 5]
```diff
@@ -130,4 +136,50 @@ public static function compare($a, $b) {
...
+        $parts['host'] = isset($parsedUrl['host']) ? $parsedUrl['host'] : '';
+        $parts['port'] = isset($parsedUrl['port']) ? ':' . $parsedUrl['port'] : '';
+        $parts['user'] = isset($parsedUrl['user']) ? $parsedUrl['user'] : '';
+        $parts['pass'] = isset($parsedUrl['pass']) ? ':' . $parsedUrl['pass'] : '';
+        $parts['pass'] = ($parts['user'] || $parts['pass']) ? $parts['pass'] . '@' : '';
+        $parts['path'] = isset($parsedUrl['path']) ? $parsedUrl['path'] : '';
+        $parts['query'] = isset($parsedUrl['query']) ? '?' . $parsedUrl['query'] : '';
+        $parts['fragment'] = isset($parsedUrl['fragment']) ? '#' . $parsedUrl['fragment'] : '';
+
+        return \implode('', $parts);
```
Comment (childish-sambino) [inline src/Twilio/Security/RequestValidator.php]: Not convinced this works properly when the URL has creds. Assume it's adapted from this: https://www.php.net/manual/en/function.parse-url.php#106731

More tests would convince me.

File: src/Twilio/Security/RequestValidator.php (hunk 2 above)
Comment (eshanholtz) [reply src/Twilio/Security/RequestValidator.php]: Was getting an "risky test" error when running `make test` since `params['bodySHA256']` is not always defined. PHP manual states that accessing an undefined key will raise an E-NOTICE level error message and return `NULL`. This line ensures the `NULL` return value while avoiding the error, since this behavior is common.

Comment (childish-sambino) [reply src/Twilio/Security/RequestValidator.php]: Just found this util: https://github.com/twilio/twilio-php/blob/master/src/Twilio/Values.php#L12

Comment (eshanholtz) [reply src/Twilio/Security/RequestValidator.php]: oh good, I'll refactor to include that.

File: src/Twilio/Security/RequestValidator.php (hunk 5 above)
Comment (eshanholtz) [reply src/Twilio/Security/RequestValidator.php]: You're right, builder adds creds to the wrong location. New tests added to validate creds in URL.
